azure-proto-core-test/UserAssignedIdentityTests.cs

```diff
@@ -0,0 +1,377 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+
+namespace azure_proto_core_test
+{
+    public class UserAssignedIdentityTests : UserAssignedIdentity
+    {
+        [TestCase(0, null, null, null, null)]
+        [TestCase(0, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98")]
+        [TestCase(1, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", null, null)]
+        [TestCase(1, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", "72f988bf-86f1-41af-91ab-2d7cd011db46", "de29bab1-49e1-4705-819b-4dfddceaaa97")]
+        [TestCase(-1, null, null, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98")]
+        [TestCase(-1, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", "72f988bf-86f1-41af-91ab-2d7cd011db48", "de29bab1-49e1-4705-819b-4dfddceaaa99")]
+        public void CompareTo(int answer, string clientId1, string principalId1, string clientId2, string principalId2)
+        {
+            UserAssignedIdentity identity1;
+            UserAssignedIdentity identity2;
+            if (clientId1 == null)
+                identity1 = new UserAssignedIdentity();
+
+            else
+                identity1 = new UserAssignedIdentity(new Guid(clientId1), new Guid(principalId1));
+
+            if (clientId2 == null)
+                identity2 = new UserAssignedIdentity();
+
+            else
+                identity2 = new UserAssignedIdentity(new Guid(clientId2), new Guid(principalId2));
+
+            Assert.AreEqual(answer, identity1.CompareTo(identity2));
+        }
+
+        [TestCase(null, null, null, null)]
+        [TestCase("72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98")]
+        public void EqualsMethodTrue(string clientId1, string principalId1, string clientId2, string principalId2)
+        {
+            UserAssignedIdentity identity1;
+            UserAssignedIdentity identity2;
+            if (clientId1 == null)
+                identity1 = new UserAssignedIdentity();
+
+            else
+                identity1 = new UserAssignedIdentity(new Guid(clientId1), new Guid(principalId1));
+
+            if (clientId2 == null)
+                identity2 = new UserAssignedIdentity();
+
+            else
+                identity2 = new UserAssignedIdentity(new Guid(clientId2), new Guid(principalId2));
+
+            Assert.IsTrue(identity1.Equals(identity2));
+        }
+
+        [TestCase(null, null, "72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98")]
+        [TestCase("72f988bf-86f1-41af-91ab-2d7cd011db47", "de29bab1-49e1-4705-819b-4dfddceaaa98", "72f988bf-86f1-41af-91ab-2d7cd011db44", "de29bab1-49e1-4705-819b-4dfddceaaa94")]
```

Comment:
same comment as above - need to exercise cases where only one of the properties differs